File: client-android/app/src/main/java/stasis/client_android/tracking/DefaultBackupTracker.kt
package stasis.client_android.tracking

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.Message
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import kotlinx.coroutines.runBlocking
import stasis.client_android.lib.collection.rules.Rule
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.SourceEntity
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.persistence.state.StateStore
import stasis.client_android.lib.tracking.BackupTracker
import stasis.client_android.lib.tracking.state.BackupState
import stasis.client_android.lib.tracking.state.serdes.BackupStateSerdes
import stasis.client_android.lib.utils.Either
import java.io.File
import java.nio.file.Path
import java.time.Duration
import java.time.Instant

class DefaultBackupTracker(
    context: Context,
    looper: Looper
) : BackupTracker, BackupTrackerView, BackupTrackerManage {
    private val trackedState: MutableLiveData<Map<OperationId, BackupState>> =
        MutableLiveData(emptyMap())

    private val handler: TrackingHandler = TrackingHandler(context, looper)

    override val state: LiveData<Map<OperationId, BackupState>> = trackedState

    override fun updates(operation: OperationId): LiveData<BackupState> =
        MediatorLiveData<BackupState>().apply {
            addSource(trackedState) { operations ->
                operations[operation]?.let { state ->
                    postValue(state)
                }
            }
        }

    override suspend fun stateOf(operation: OperationId): BackupState? {
        return trackedState.value?.get(operation)
    }

    override fun remove(operation: OperationId) {
        handler.obtainMessage().let { msg ->
            msg.obj = TrackerEvent.RemoveOperation(operation)
            handler.sendMessage(msg)
        }
    }

    override fun clear() {
        handler.obtainMessage().let { msg ->
            msg.obj = TrackerEvent.ClearOperations
            handler.sendMessage(msg)
        }
    }

    override fun started(operation: OperationId, definition: DatasetDefinitionId) = send(
        event = BackupEvent.Started(
            operation = operation,
            definition = definition
        )
    )

    override fun entityDiscovered(operation: OperationId, entity: Path) = send(
        event = BackupEvent.EntityDiscovered(
            operation = operation,
            entity = entity
        )
    )


    override fun specificationProcessed(
        operation: OperationId,
        unmatched: List<Pair<Rule, Throwable>>
    ) {
        if (unmatched.isNotEmpty()) {
            send(
                event = BackupEvent.SpecificationProcessed(
                    operation = operation,
                    unmatched = unmatched.map { (rule, e) ->
                        "Rule [${rule.asString()}] failed with [${e.message}]"
                    }
                )
            )
        }
    }

    override fun entityExamined(operation: OperationId, entity: Path) = send(
        event = BackupEvent.EntityExamined(
            operation = operation,
            entity = entity
        )
    )

    override fun entitySkipped(operation: OperationId, entity: Path) = send(
        event = BackupEvent.EntitySkipped(
            operation = operation,
            entity = entity
        )
    )

    override fun entityCollected(operation: OperationId, entity: SourceEntity) = send(
        event = BackupEvent.EntityCollected(
            operation = operation,
            entity = entity
        )
    )

    override fun entityProcessingStarted(operation: OperationId, entity: Path, expectedParts: Int) = send(
        event = BackupEvent.EntityProcessingStarted(
            operation = operation,
            entity = entity,
            expectedParts = expectedParts
        )
    )

    override fun entityPartProcessed(operation: OperationId, entity: Path) = send(
        event = BackupEvent.EntityPartProcessed(
            operation = operation,
            entity = entity
        )
    )

    override fun entityProcessed(
        operation: OperationId,
        entity: Path,
        metadata: Either<EntityMetadata, EntityMetadata>
    ) = send(
        event = BackupEvent.EntityProcessed(
            operation = operation,
            entity = entity,
            metadata = metadata
        )
    )


    override fun metadataCollected(operation: OperationId) = send(
        event = BackupEvent.MetadataCollected(
            operation = operation
        )
    )

    override fun metadataPushed(operation: OperationId, entry: DatasetEntryId) = send(
        event = BackupEvent.MetadataPushed(
            operation = operation
        )
    )

    override fun failureEncountered(operation: OperationId, entity: Path, failure: Throwable) = send(
        event = BackupEvent.EntityFailed(
            operation = operation,
            entity = entity,
            reason = failure
        )
    )

    override fun failureEncountered(operation: OperationId, failure: Throwable) = send(
        event = BackupEvent.FailureEncountered(
            operation = operation,
            reason = failure
        )
    )

    override fun completed(operation: OperationId) = send(
        event = BackupEvent.Completed(
            operation = operation
        )
    )

    private fun send(event: BackupEvent) {
        handler.obtainMessage().let { msg ->
            msg.obj = event
            handler.sendMessage(msg)
        }
    }

    private inner class TrackingHandler(context: Context, looper: Looper) : Handler(looper) {
        private var _state: Map<OperationId, BackupState> = emptyMap()
        private var updates: Int = 0
        private var persistScheduled: Boolean = false

        private val store: StateStore<Map<OperationId, BackupState>> = StateStore(
            target = File(context.filesDir, "state/backups").toPath(),
            serdes = BackupStateSerdes
        )

        init {
            obtainMessage().let { msg ->
                msg.obj = TrackerEvent.RestoreState
                sendMessage(msg)
            }
        }

        override fun handleMessage(msg: Message) {
            when (val event = msg.obj) {
                is BackupEvent -> {
                    val existing = if (event is BackupEvent.Started) {
                        BackupState.start(event.operation, definition = event.definition)
                    } else {
                        val existing = _state[event.operation]

                        require(existing != null) {
                            "Expected existing state for operation [${event.operation}] but none was found"
                        }

                        existing
                    }

                    val updated = when (event) {
                        is BackupEvent.EntityDiscovered -> existing.entityDiscovered(event.entity)
                        is BackupEvent.SpecificationProcessed -> existing.specificationProcessed(event.unmatched)
                        is BackupEvent.EntityExamined -> existing.entityExamined(event.entity)
                        is BackupEvent.EntitySkipped -> existing.entitySkipped(event.entity)
                        is BackupEvent.EntityCollected -> existing.entityCollected(event.entity)
                        is BackupEvent.EntityProcessingStarted -> existing.entityProcessingStarted(
                            event.entity,
                            event.expectedParts
                        )
                        is BackupEvent.EntityPartProcessed -> existing.entityPartProcessed(event.entity)
                        is BackupEvent.EntityProcessed -> existing.entityProcessed(event.entity, event.metadata)
                        is BackupEvent.EntityFailed -> existing.entityFailed(event.entity, event.reason)
                        is BackupEvent.FailureEncountered -> existing.failureEncountered(event.reason)
                        is BackupEvent.MetadataCollected -> existing.backupMetadataCollected()
                        is BackupEvent.MetadataPushed -> existing.backupMetadataPushed()
                        is BackupEvent.Completed -> {
                            obtainMessage().let { message ->
                                message.obj = TrackerEvent.PersistState
                                sendMessage(message)
                            }
                            existing.backupCompleted()
                        }
                        else -> existing
                    }

                    updates += 1

                    if (updates >= PersistAfterEvents) {
                        obtainMessage().let { message ->
                            message.obj = TrackerEvent.PersistState
                            sendMessageAtFrontOfQueue(message)
                        }
                    }

                    if (!persistScheduled) {
                        persistScheduled = true
                        obtainMessage().let { message ->
                            message.obj = TrackerEvent.PersistState
                            sendMessageDelayed(message, PersistAfterPeriod.toMillis())
                        }
                    }

                    val now = Instant.now()
                    val filtered = _state.filterValues { it.started.plusMillis(MaxRetention.toMillis()).isAfter(now) }

                    _state = filtered + (event.operation to updated)
                    trackedState.postValue(_state)
                }

                is TrackerEvent -> {
                    when (event) {
                        is TrackerEvent.RestoreState -> runBlocking {
                            when (val state = store.restore()) {
                                null -> Unit // do nothing
                                else -> {
                                    _state = state
                                    trackedState.postValue(_state)
                                }
                            }
                        }

                        is TrackerEvent.PersistState -> runBlocking {
                            updates = 0
                            persistScheduled = false
                            store.persist(state = _state)
                        }

                        is TrackerEvent.RemoveOperation -> {
                            _state = _state - event.operation
                            trackedState.postValue(_state)

                            obtainMessage().let { message ->
                                message.obj = TrackerEvent.PersistState
                                sendMessage(message)
                            }
                        }

                        is TrackerEvent.ClearOperations -> {
                            _state = emptyMap()
                            trackedState.postValue(_state)

                            obtainMessage().let { message ->
                                message.obj = TrackerEvent.PersistState
                                sendMessage(message)
                            }
                        }
                    }
                }

                else -> throw IllegalArgumentException("Unexpected message encountered: [$event]")
            }
        }
    }

    private sealed class TrackerEvent {
        data object RestoreState : TrackerEvent()
        data object PersistState : TrackerEvent()
        data class RemoveOperation(val operation: OperationId) : TrackerEvent()
        data object ClearOperations : TrackerEvent()
    }

    private sealed class BackupEvent {
        abstract val operation: OperationId

        data class Started(
            override val operation: OperationId,
            val definition: DatasetDefinitionId
        ) : BackupEvent()

        data class EntityDiscovered(
            override val operation: OperationId,
            val entity: Path
        ) : BackupEvent()

        data class SpecificationProcessed(
            override val operation: OperationId,
            val unmatched: List<String>
        ) : BackupEvent()

        data class EntityExamined(
            override val operation: OperationId,
            val entity: Path
        ) : BackupEvent()

        data class EntitySkipped(
            override val operation: OperationId,
            val entity: Path
        ) : BackupEvent()

        data class EntityCollected(
            override val operation: OperationId,
            val entity: SourceEntity
        ) : BackupEvent()

        data class EntityProcessingStarted(
            override val operation: OperationId,
            val entity: Path,
            val expectedParts: Int
        ) : BackupEvent()

        data class EntityPartProcessed(
            override val operation: OperationId,
            val entity: Path
        ) : BackupEvent()

        data class EntityProcessed(
            override val operation: OperationId,
            val entity: Path,
            val metadata: Either<EntityMetadata, EntityMetadata>
        ) : BackupEvent()

        data class EntityFailed(
            override val operation: OperationId,
            val entity: Path,
            val reason: Throwable
        ) : BackupEvent()

        data class FailureEncountered(
            override val operation: OperationId,
            val reason: Throwable
        ) : BackupEvent()

        data class MetadataCollected(
            override val operation: OperationId
        ) : BackupEvent()

        data class MetadataPushed(
            override val operation: OperationId
        ) : BackupEvent()

        data class Completed(
            override val operation: OperationId
        ) : BackupEvent()
    }

    companion object {
        private val MaxRetention: Duration = Duration.ofDays(30)
        private const val PersistAfterEvents: Int = 1000
        private val PersistAfterPeriod: Duration = Duration.ofSeconds(30)
    }
}

File: client/src/main/scala/stasis/client/tracking/trackers/DefaultBackupTracker.scala
package stasis.client.tracking.trackers

import java.nio.file.Path
import java.time.Instant

import scala.concurrent.duration.FiniteDuration
import scala.concurrent.ExecutionContext
import scala.concurrent.Future

import org.apache.pekko.NotUsed
import org.apache.pekko.actor.typed.scaladsl.LoggerOps
import org.apache.pekko.stream.scaladsl.Source
import org.slf4j.Logger
import org.slf4j.LoggerFactory

import stasis.client.collection.rules.Rule
import stasis.client.model.EntityMetadata
import stasis.client.model.SourceEntity
import stasis.client.tracking.BackupTracker
import stasis.client.tracking.state.BackupState
import stasis.client.tracking.trackers.DefaultBackupTracker.updateState
import stasis.core.persistence.backends.EventLogBackend
import stasis.core.persistence.events.EventLog
import io.github.sndnv.layers.streaming.Operators.ExtendedSource
import stasis.shared.model.datasets.DatasetDefinition
import stasis.shared.model.datasets.DatasetEntry
import stasis.shared.ops.Operation

class DefaultBackupTracker(
  maxRetention: FiniteDuration,
  backend: EventLogBackend[DefaultBackupTracker.BackupEvent, Map[Operation.Id, BackupState]]
)(implicit ec: ExecutionContext)
    extends BackupTracker {
  import DefaultBackupTracker.BackupEvent

  private val log: Logger = LoggerFactory.getLogger(this.getClass.getName)

  private val events: EventLog[BackupEvent, Map[Operation.Id, BackupState]] = EventLog(
    backend = backend,
    updateState = { case (event, state) => updateState(event, state, maxRetention) }
  )

  override def state: Future[Map[Operation.Id, BackupState]] = events.state

  override def updates(operation: Operation.Id): Source[BackupState, NotUsed] = {
    val latest = Source.future(events.state).map(_.get(operation)).collect { case Some(state) => state }
    latest.concat(events.stateStream.dropLatestDuplicates(_.get(operation)))
  }

  override def exists(operation: Operation.Id): Future[Boolean] =
    events.state.map(_.contains(operation))

  override def remove(operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Operation removed", operation)

    val _ = events.store(event = BackupEvent.Removed(operation))
  }

  override def started(
    definition: DatasetDefinition.Id
  )(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Operation started", operation)

    val _ = events.store(event = BackupEvent.Started(operation, definition))
  }

  override def entityDiscovered(
    entity: Path
  )(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Entity [{}] discovered", operation, entity)

    val _ = events.store(event = BackupEvent.EntityDiscovered(operation, entity))
  }

  override def specificationProcessed(
    unmatched: Seq[(Rule, Throwable)]
  )(implicit operation: Operation.Id): Unit =
    if (unmatched.isEmpty) {
      log.debugN("[{}] (backup) - Specification successfully processed", operation)
    } else {
      val failures = unmatched.map { case (rule, e) =>
        s"Rule [${rule.asString}] failed with [${e.getClass.getSimpleName} - ${e.getMessage}]"
      }

      log.debugN(
        "[{}] (backup) - Specification processed with [{}] unmatched rules: [{}]",
        operation,
        failures.length,
        failures.mkString("\n\t", "\n\t", "\n")
      )

      val _ = events.store(event = BackupEvent.SpecificationProcessed(operation, failures))
    }

  override def entityExamined(
    entity: Path,
    metadataChanged: Boolean,
    contentChanged: Boolean
  )(implicit operation: Operation.Id): Unit = {
    log.debugN(
      "[{}] (backup) - Entity [{}] examined; metadata changed: [{}]; content changed: [{}]",
      operation,
      entity,
      metadataChanged,
      contentChanged
    )

    val _ = events.store(event = BackupEvent.EntityExamined(operation, entity))
  }

  override def entitySkipped(
    entity: Path
  )(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Entity [{}] skipped", operation, entity)

    val _ = events.store(event = BackupEvent.EntitySkipped(operation, entity))
  }

  override def entityCollected(
    entity: SourceEntity
  )(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Entity [{}] collected", operation, entity.path)

    val _ = events.store(event = BackupEvent.EntityCollected(operation, entity))
  }

  override def entityProcessingStarted(
    entity: Path,
    expectedParts: Int
  )(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Entity [{}] processing started; expected parts: [{}]", operation, entity, expectedParts)

    val _ = events.store(event = BackupEvent.EntityProcessingStarted(operation, entity, expectedParts))
  }

  override def entityPartProcessed(
    entity: Path
  )(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Part for entity [{}] processed", operation, entity)

    val _ = events.store(event = BackupEvent.EntityPartProcessed(operation, entity))
  }

  override def entityProcessed(
    entity: Path,
    metadata: Either[EntityMetadata, EntityMetadata]
  )(implicit operation: Operation.Id): Unit = {
    val contentChanged = metadata.isLeft
    log.debugN("[{}] (backup) - Entity [{}] processed; content changed: [{}]", operation, entity, contentChanged)

    val _ = events.store(event = BackupEvent.EntityProcessed(operation, entity, metadata))
  }

  override def metadataCollected()(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Metadata collected", operation)

    val _ = events.store(event = BackupEvent.MetadataCollected(operation))
  }

  override def metadataPushed(
    entry: DatasetEntry.Id
  )(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Metadata pushed; entry [{}] created", operation, entry)

    val _ = events.store(event = BackupEvent.MetadataPushed(operation))
  }

  override def failureEncountered(
    failure: Throwable
  )(implicit operation: Operation.Id): Unit = {
    log.debugN(
      "[{}] (backup) - Failure encountered: [{} - {}]",
      operation,
      failure.getClass.getSimpleName,
      failure.getMessage
    )

    val _ = events.store(event = BackupEvent.FailureEncountered(operation, failure))
  }

  override def failureEncountered(
    entity: Path,
    failure: Throwable
  )(implicit operation: Operation.Id): Unit = {
    log.debugN(
      "[{}] (backup) - Failure encountered while processing entity [{}]: [{} - {}]",
      operation,
      entity,
      failure.getClass.getSimpleName,
      failure.getMessage
    )

    val _ = events.store(event = BackupEvent.EntityFailed(operation, entity, failure))
  }

  override def completed()(implicit operation: Operation.Id): Unit = {
    log.debugN("[{}] (backup) - Operation completed", operation)

    val _ = events.store(event = BackupEvent.Completed(operation))
  }
}

object DefaultBackupTracker {
  def apply(
    maxRetention: FiniteDuration,
    createBackend: Map[Operation.Id, BackupState] => EventLogBackend[BackupEvent, Map[Operation.Id, BackupState]]
  )(implicit ec: ExecutionContext): DefaultBackupTracker =
    new DefaultBackupTracker(backend = createBackend(Map.empty), maxRetention = maxRetention)

  sealed trait BackupEvent {
    def operation: Operation.Id
  }

  object BackupEvent {
    final case class Started(
      override val operation: Operation.Id,
      definition: DatasetDefinition.Id
    ) extends BackupEvent

    final case class EntityDiscovered(
      override val operation: Operation.Id,
      entity: Path
    ) extends BackupEvent

    final case class SpecificationProcessed(
      override val operation: Operation.Id,
      unmatched: Seq[String]
    ) extends BackupEvent

    final case class EntityExamined(
      override val operation: Operation.Id,
      entity: Path
    ) extends BackupEvent

    final case class EntitySkipped(
      override val operation: Operation.Id,
      entity: Path
    ) extends BackupEvent

    final case class EntityCollected(
      override val operation: Operation.Id,
      entity: SourceEntity
    ) extends BackupEvent

    final case class EntityProcessingStarted(
      override val operation: Operation.Id,
      entity: Path,
      expectedParts: Int
    ) extends BackupEvent

    final case class EntityPartProcessed(
      override val operation: Operation.Id,
      entity: Path
    ) extends BackupEvent

    final case class EntityProcessed(
      override val operation: Operation.Id,
      entity: Path,
      metadata: Either[EntityMetadata, EntityMetadata]
    ) extends BackupEvent

    final case class EntityFailed(
      override val operation: Operation.Id,
      entity: Path,
      reason: Throwable
    ) extends BackupEvent

    final case class FailureEncountered(
      override val operation: Operation.Id,
      reason: Throwable
    ) extends BackupEvent

    final case class MetadataCollected(
      override val operation: Operation.Id
    ) extends BackupEvent

    final case class MetadataPushed(
      override val operation: Operation.Id
    ) extends BackupEvent

    final case class Completed(
      override val operation: Operation.Id
    ) extends BackupEvent

    final case class Removed(
      override val operation: Operation.Id
    ) extends BackupEvent
  }

  def updateState(
    event: BackupEvent,
    state: Map[Operation.Id, BackupState],
    maxRetention: FiniteDuration
  ): Map[Operation.Id, BackupState] = {
    import BackupEvent._

    val updated = event match {
      case Started(_, definition) =>
        BackupState.start(event.operation, definition)

      case _ =>
        require(
          state.contains(event.operation),
          s"Expected existing state for operation [${event.operation.toString}] but none was found"
        )

        val existing = state(event.operation)

        event match {
          case EntityDiscovered(_, entity)                       => existing.entityDiscovered(entity)
          case SpecificationProcessed(_, unmatched)              => existing.specificationProcessed(unmatched)
          case EntityExamined(_, entity)                         => existing.entityExamined(entity)
          case EntitySkipped(_, entity)                          => existing.entitySkipped(entity)
          case EntityCollected(_, entity)                        => existing.entityCollected(entity)
          case EntityProcessingStarted(_, entity, expectedParts) => existing.entityProcessingStarted(entity, expectedParts)
          case EntityPartProcessed(_, entity)                    => existing.entityPartProcessed(entity)
          case EntityProcessed(_, entity, metadata)              => existing.entityProcessed(entity, metadata)
          case EntityFailed(_, entity, reason)                   => existing.entityFailed(entity, reason)
          case FailureEncountered(_, e)                          => existing.failureEncountered(e)
          case _: MetadataCollected                              => existing.backupMetadataCollected()
          case _: MetadataPushed                                 => existing.backupMetadataPushed()
          case _: Completed                                      => existing.backupCompleted()
          case _                                                 => existing
        }
    }

    val now = Instant.now()
    val filtered = state.filter(_._2.started.plusMillis(maxRetention.toMillis).isAfter(now))

    event match {
      case Removed(operation) => filtered - operation
      case _                  => filtered + (event.operation -> updated)
    }
  }
}
